man/taunet-echo.man
.\" Copyright (c) 2015 Bart Massey
.\" This work is licensed under the GPL version 3 or later.
.\" Please see the file COPYING in the source
.\" distribution of this software for license terms.
.TH TAUNET-ECHO 1 2015-11-28 "TauNet" "TauNet Manual"
.SH NAME
taunet-echo \- TauNet echo server
.SH SYNOPSIS
.B taunet-echo
.RI [ -p ]
.RI [ -d ]
.RI [ -f
.BR <fail-user> ]
.SH DESCRIPTION
.PP
This server will listen for a TauNet message and echo it
back as the body of a message sent to the TauNet server
presumptively at the source address of the original.
.PP
When sending to a local address no reply will be sent, to
avoid messaging loops.
.PP
The return message body will consist of a line reporting the
date and time that the request was received and the
requesting IP address, followed by the request message in
its entirety, but truncated to the maximum message length.
If the incoming message is ill-formatted, the return message
body will start with an error report.
.PP
The return message will be listed as from a TauNet user
specifiable on the command line, and will be addressed to
the incoming sender if possible.
.PP
For encryption, a file
.B key.txt
should be present in the startup directory, containing
a single line comprising an ASCII CipherSaber-2 key.
.SH OPTIONS
.TP
.B -p
Act as a plaintext server, that is, do not do CipherSaber-2
encryption of input or output.
.TP
.B -d
Do unspecified debugging things.
.TP
.BR "-f " fail-user
Specify a TauNet username that will be the used to form the
"from" and "to" name for failure messages, by appending
"-FROM" and "-TO" respectively. Default is "FAIL".
.SH COMMANDS
.PP
The echo server will process lines beginning with @ at the
beginning of the incoming message body as commands. These
commands allow the client to modify the servers echo
behavior. When the same command is processed multiple times,
the last value seen is used.
.TP
.BI @delay " <secs>"
Makes the server delay for a fractional number of seconds
before replying. The default delay is 0.1. The maximum delay
is 60.
.TP
.BI @replies " <count>"
Makes the server repeatedly send the reply. The default
count is 1. The maximum count is 10. Zero is a legitimate
count. The delay happens before each repetition.
.SH SEE ALSO
taunet-send(1), taunet-node(1)
.SH AUTHOR
Bart Massey
.I <bart@cs.pdx.edu>
.SH LICENSE
This work is licensed under the GPL version 3 or later.
Please see the file COPYING in the source distribution of
this software for license terms.


taunet-echo.hs
{-# LANGUAGE OverloadedStrings, BangPatterns #-}
-- Copyright © 2015 Bart Massey
-- [This program is licensed under the GPL version 3 or later.]
-- Please see the file COPYING in the source
-- distribution of this software for license terms.

-- TauNet echo server

import Control.Concurrent.Thread.Delay
import Control.Monad
import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as BSC
import Network.Socket hiding (send, sendTo, recv, recvFrom)
import System.Console.ParseArgs
import System.Exit
import System.Posix.Process
import Text.Printf

import DateTime
import LocalAddr
import TaunetUtil
import TaunetMessage

reapChildren :: IO ()
reapChildren = do
  maybeChild <- getAnyProcessStatus False True
  case maybeChild of
    Nothing -> return ()
    Just (pid, status) -> do
      when (status /= Exited ExitSuccess) $
           logString $ printf "%s exited prematurely: %s"
                         (show pid) (show status)
      reapChildren

data ArgIndex = ArgPlain | ArgDebug | ArgFailUser
              deriving (Eq, Ord, Enum, Show)

argd :: [ Arg ArgIndex ]
argd = [ Arg {
           argIndex = ArgPlain,
           argAbbr = Just 'p',
           argName = Just "plain",
           argData = Nothing,
           argDesc = "Do not encrypt the incoming or outgoing messages." },
         Arg {
           argIndex = ArgDebug,
           argAbbr = Just 'd',
           argName = Just "debug",
           argData = Nothing,
           argDesc = "Do debugging things." },
         Arg {
           argIndex = ArgFailUser,
           argDesc = "Specify TauNet username used in failure reports.",
           argAbbr = Just 'f',
           argName = Just "fail-user",
           argData = argDataDefaulted "username-prefix"
                     ArgtypeString "FAIL" } ]

main :: IO ()
main = do
  argv <- parseArgsIO ArgsComplete argd
  let !encrypted = not $ gotArg argv ArgPlain
  let !debug = gotArg argv ArgDebug
  let !failUser = getRequiredArg argv ArgFailUser
  maybeKey <- maybeGetKey encrypted
  taunetSocket <- socket AF_INET Stream defaultProtocol
  setSocketOption taunetSocket ReuseAddr 1
  bind taunetSocket $ SockAddrInet taunetPort 0
  listen taunetSocket 1
  logString "taunet-echo begins"
  forever $ do
    (recvSocket, recvAddr) <- accept taunetSocket
    _ <- forkProcess $ do
      close taunetSocket
      let ha = hostAddr recvAddr
      recvTime <- getTimeRFC3339
      plaintext <- receiveMessage
                     (Just (2 * maxMessageSize))
                     maybeKey
                     recvSocket
      when (BS.length plaintext == 0) $ do
        logString $ printf
                      "(%s) ping (zero-length) message received and discarded"
                      (show ha)
        exitSuccess
      let received = parseMessage True plaintext
      let knobs = getKnobs received
      when debug $ do print received
      logMessage ha received
      localAddresses <- getLocalAddresses
      when (ha `elem` localAddresses) exitSuccess
      let replyMessage =
              case received of
                Right message -> swapFromTo message
                Left failure ->
                    Message {
                      messageTo = failUser ++ "-TO",
                      messageFrom = failUser ++ "-FROM",
                      messageCommands = [],
                      messageBody =
                          BSC.pack (failureMessage failure) +++
                                   "\r\n" +++ failureBody failure }
              where
                swapFromTo m@(Message { messageTo = t, messageFrom = f }) =
                    m { messageTo = f, messageFrom = t }
      replicateM_ (knobReplies knobs) $ do
        delay $ floor $ 1000000 * knobDelay knobs
        result <- generateMessage maybeKey
                                  (Just (recvTime, recvAddr))
                                  (hostAddr recvAddr)
                                  replyMessage
        case result of
          Left msg -> logString $ "response failed: " ++ msg
          Right () -> return ()
      exitSuccess
    reapChildren
